images/dnsmasq/validate_dynamic_deps.py

```diff
@@ -0,0 +1,170 @@
+#/usr/bin/env python3
+
+# Copyright 2022 The Kubernetes Authors.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# This script requires python 3.8+
+import os
+import shutil
+import subprocess
+import pathlib
+import sys
+from typing import Optional
+
+IMAGE_DIR = '.test-images'
+
+
+def replace_img_special_chars(image_ref: str) -> str:
+    """Replace container image name characters that cannot be a part of container names or file names."""
+    return image_ref.lower().replace('/', '_').replace(':', '_')
+
+
+def relative_to_root(path: pathlib.Path) -> pathlib.Path:
+    return path.relative_to(path.anchor)
+
+
+def detect_architecture(cwd: pathlib.Path, target_binary: pathlib.Path) -> str:
+    """Read the architecture of the ELF file target_binary using the file command"""
+    result = subprocess.run(['file', target_binary],
+                            capture_output=True,
+                            text=True,
+                            cwd=cwd)
+    return result.stdout.split(",")[1].strip()
+
+
+def untar_container_image(image_ref: str, output_dir: pathlib.Path):
+    """Create a docker container using image_ref use tar to extract the filesystem to output_dir."""
+    container_name = replace_img_special_chars(image_ref)
+    subprocess.run(['docker', 'create', '--name', container_name, image_ref],
+                   capture_output=True)
+    subprocess.run(f'docker export {container_name} | tar x',
+                   shell=True,
+                   cwd=output_dir,
+                   capture_output=True)
+    subprocess.run(['docker', 'rm', container_name], capture_output=True)
+
+
+def detect_dependencies(cwd: pathlib.Path,
+                        target_binary: pathlib.Path) -> [str]:
+    """Use objdump to read the ELF headers for target_binary and return a list of .so files."""
+    result = subprocess.run(['objdump', '--private-headers', target_binary],
```

Comment:
Why `objdump` is used?
IIUC `ldd target_binary` would give you a full dependency graph with filepath of those dependencies

Reply:
`objdump` is used because it reads the binary's. `ldd` will actually try to run the binary using a special `LD_TRACE_LOADED_OBJECTS` flag which gets the dependencies.

Running ldd on my x86 machine against the x86 container
```shell
❯ ldd arm-container-test/k8s-dns-dnsmasq-amd64-1.22.3/usr/sbin/dnsmasq
        linux-vdso.so.1 (0x00007ffc3b9f5000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb2f44f8000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb2f4748000)
```

Running ldd on my x86 machine against the ARM container

```shell
❯ ldd arm-container-test/k8s-dns-dnsmasq-amd64-1.22.3/usr/sbin/dnsmasq
        linux-vdso.so.1 (0x00007ffc3b9f5000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb2f44f8000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb2f4748000)
```

Reply:
I added a comment to the function explaining why it uses objdump